loop event repeater

dr7ana opened the pull request:
- Repeating events can be instantiated using `::call_every(...)`
- `EventHandler` object is added for lifetime management of repeated events. Application can start/pause/stop at will
- Application has the choice of tying the lifetime of the event to the calling object, or using the `EventHandler`
- 002 test case added sending via repeater
- `std::move` -> `std::forward` where necessary

dr7ana committed:
loop event repeater

- Repeating events can be instantiated using ::call_every(...)
- 002 test case added sending via repeater
- std::move -> std::forward where necessary

dr7ana committed:
Loop EventHandler added, lifetime management

- EventHandler object is added for lifetime management of repeated events. Application can start/pause/stop at will
- Application has the choice of tying the lifetime of the event to the calling object, or using the EventHandler

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -124,6 +145,23 @@ namespace oxen::quic
         }
 
+        template <typename Callable>
+        void call_every(loop_time interval, std::weak_ptr<void> caller, Callable&& f)
```
This interface feels a bit awkward: it seems to be catering to a specific case where I want my timer to be tied to a specific shared_ptr, which means to manage my own timer when it should be tied to something other than a shared_ptr I still get forced into using an external shared ptr unrelated to the timer, so I'd have to do something like:

```C++
struct Dummy {};
auto x = std::make_shared_ptr<Dummy>();
loop->call_every(250ms, x, [] { task(); });

// ...
x.reset();
```
to manage it.  There's also no way I can say "repeat this thing forever" without needing to create and store an unnecessary shared_ptr essentially forever (and even then there's likely a segfault if my shared_ptr outlives the loop), and there's no ability to pause a timer for a while and resume it later.

I do realize this is coming out of Lokinet, but we can do better than simply copy Lokinet's implementation that is catered for Lokinet's specific use case.  (It's actually decently fine for Lokinet's specific use case as is, but a generic library interface needs to be more general).

A possible alternative could work like this:

```C++
// Don't care about controlling it:
loop->call_every(250ms, [] { task(); });

// Option 1: return an ID and add ID-taking control functions to Loop:
auto timer_id = loop->call_every(1s, [] { task(); });
loop->stop_timer(timer_id);
loop->start_timer(timer_id, 500ms);

// Option 2: return an object that manages it:
auto timer = loop->call_every(1s, [] { task(); });
timer->stop();
timer->start(500ms);


// Possibilities for cancelling or pausing a timer from within the timer callback:
class timer_cancel : public std::runtime_error { timer_cancel() : std::runtime_error{"timer cancelled"} {} };
class timer_stop : public std::runtime_error { timer_stop() : std::runtime_error{"timer stopped"} {} };
void task() {
    auto p = some_weak_ptr.lock();
    if (!x)
        throw timer_cancel{};
    //
    // do stuff
    //
    if (++num_calls == 300) {
        num_calls = 0;
        throw timer_stop{};
    }
}
```

dr7ana replied:
I see your point, but creating (option 1) an interface by which loop events are manipulated using their uuids that also need to be implemented or (option 2) an entire timer object is way beyond the scope of what this is meant to effect. 

Tying the timer to the lifetime of an object also makes perfect sense when considering that the logic being repeatedly invoked is also tied to it's lifetime. The test case I added to `002` makes this clear, as the timer repeatedly calls on the stream to send a message

jagerman replied:
> creating (option 1) an interface by which loop events are manipulated using their uuids that also need to be implemented or (option 2) an entire timer object is way beyond the scope of what this is meant to effect.

Then drop this PR and keep this logic lokinet side.  If we put repeatable events into libquic's event loop then it should be a well-designed interface, and not just copied from the one that Lokinet had.  This implementation in Lokinet was designed when the underlying event code was designed to be swappable to different underlying event loops with potentially different capabilities, which is partly why it has this shared_ptr lifetime management.  It was also good enough for what lokinet needed, but never intended to be a generic event loop timer interface.

> Tying the timer to the lifetime of an object also makes perfect sense when considering that the logic being repeatedly invoked is also tied to it's lifetime. The test case I added to 002 makes this clear, as the timer repeatedly calls on the stream to send a message

Yes, sometimes it does, and the 002 test case is tailored to that use case.  But that's not the only use case for a timer, and a generic library interface has a higher bar in being usable across different use cases.

jagerman commented on src/loop.cpp:
```diff
@@ -29,6 +29,44 @@ namespace oxen::quic
+                [](evutil_socket_t, short, void* s) {
+                    auto* self = reinterpret_cast<EventHandler*>(s);
+                    // execute callback
+                    self->f();
```
We're in a C callback here: we cannot allow exceptions to bubble up through here.

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -36,6 +53,18 @@ namespace oxen::quic
+            h.start(loop(), delay, [hdnlr = std::move(handler), func = std::forward<Callable>(hook)]() mutable {
+                func();
+                hdnlr.reset();
+            });
```
This feels unnecessarily complex when libevent has the ability to schedule a delayed, non-repeating event.

dr7ana replied:
valid. I added a boolean check to either set the `EV_PERSIST` flag or no flag, which will schedule a one-time non-repeating event

dr7ana committed:
Logic modifications:

- Cleaning of old tickers on new ticker creation and on loop close
- `std::invocable` where appropriate to constrain functions
- [[nodiscard]] and `std::move` -> `std::forward` where also appropriate

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -22,6 +22,45 @@ namespace oxen::quic
 
     static void setup_libevent_logging();
 
+    using loop_time = std::chrono::microseconds;
```
Ditch this typedef: the original is a more descriptive type to have in the exposed API than the typedef.

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -22,6 +22,45 @@ namespace oxen::quic
+
+        bool start();
+        bool pause();
+        bool stop();
```
`stop()` seems unnecessary: its implementation is effectively the same as the destructor, which means an object on which you have called `stop()` is a sort of unusable zombie object.  It would be preferable to use the RAII pattern and have the destructor be the (only) permanent stop.  Then `pause()` can get renamed to `stop()` (which is a nicer linguistic inverse of `start()`), and you simplify the state tracking to just one atomic bool.

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -36,6 +75,22 @@ namespace oxen::quic
         std::mutex job_queue_mutex;
 
+        template <typename Callable>
+        void add_oneshot_event(loop_time delay, Callable&& hook)
```
There's little point here in taking `hook` as a forwarding reference, because you are always capturing it here in the lambda, and so the `Callable&&` followed by the `func = std::forward<Callable>(hook)` capture just delays but doesn't eliminate the need to move or copy -- you still have to move or copy when you construct the lambda.  Thus you can simplify it all by copying or moving at the call site by making this take a non-forwarding reference (and you can also use a concept to make sure its callable with 0 arguments):

```C++
template <std::invocable Callback>
void f(Callback cb) {
    [cb=std::move(cb)] { ... }
}
```

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -36,6 +75,22 @@ namespace oxen::quic
+                    delay,
+                    [hdnlr = std::move(handler), func = std::forward<Callable>(hook)]() mutable {
+                        func();
+                        hdnlr.reset();
```
This circular ownership introduces some difficulties:
1. if `func()` throws this leaks because `.reset()` won't get called.
2. if the event loop stops, this leaks.

1. is relatively easy to solve: just steal the shared_ptr before the call the func():
```C++
auto h = std::move(hdnlr);
func();
```
which will always release the shared pointer regardless of how this leaves the lambda.

2. is more difficult.  One solution would be for the Loop object to keep a `std::forward_list<std::weak_ptr<EventHandler>>`, and then:
- just after the loop stops, you walk the list and assign nullptr to the `.f` of any objects that `lock()` successfully.
- each time someone constructs a new EventHandler, you walk the list and remove any weak pointers that have `.expired()` (to keep this from growing indefinitely).

Another approach is what oxenmq does where the EventHandler only lives in the loop and you get some sort of atomic ID to reference it.  (But I think I prefer the weak pointer approach).

dr7ana replied:
Yeah good thinking, thanks for pointing this out

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -58,7 +115,9 @@ namespace oxen::quic
         {
-            return [this, f = std::move(f)](T* ptr) { return call_get([f = std::move(f), ptr]() { return f(ptr); }); };
+            return [this, func = std::forward<Callable>(f)](T* ptr) {
+                return call_get([f = std::forward<Callable>(func), ptr]() { return f(ptr); });
```
See previous comment about the *outer* std::forward here, and about taking a `std::invocable<T*> Callable` by value rather than forwarding reference.

The *inner* std::forward here is actually harmful: if `Callable` is a const lvalue reference then the inner forward here gives you a const value reference to `func`, which forces func to be copied rather than moved into `f`, which definitely isn't what you want.

The outer lambda here should also be `mutable` so that it's `func` can be effectively moved.  (The previous implementation was also copying, because the std::move yields a *const* rvalue reference without `mutable`).

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -77,7 +136,7 @@ namespace oxen::quic
         std::shared_ptr<T> shared_ptr(T* obj, Callable&& deleter)
         {
-            return std::shared_ptr<T>(obj, wrapped_deleter<T>(std::move(deleter)));
+            return std::shared_ptr<T>(obj, wrapped_deleter<T>(std::forward<Callable>(deleter)));
```
This one is good; it's full on forwarding the argument exactly as provided to another function.

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -22,6 +22,45 @@ namespace oxen::quic
+        bool is_stopped() const { return _is_stopped; }
+
+        bool start();
+        bool pause();
```
These need documenting comments on how they are meant to be used

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -124,7 +183,75 @@ namespace oxen::quic
+        */
+        template <typename Callable>
+        [[nodiscard]] std::shared_ptr<EventHandler> call_every(
+                loop_time interval, Callable&& f, bool start_immediately = true)
```
I think `start_immediately` should be renamed to just `start`, or perhaps inverted to `bool delay_start = false`: the `start_immediately` name made me guess (until I read the code) that this might be doing a sort of "fire now and then repeat every interval".

dr7ana replied:
This feels nit-picky? It's pretty clear what it does, and the comment block explains it explicitly above

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -124,7 +183,75 @@ namespace oxen::quic
+                call_soon([this,
+                           func = std::forward<Callable>(hook),
+                           target_time = get_timestamp<loop_time>() + delay]() mutable {
+                    auto updated_delay = target_time - get_timestamp<loop_time>();
```
`get_time` is the function you want here rather than `get_timestamp` for better type safety.  (`get_timestamp()` is really only there for feeding integral timestamps into ngtcp2 functions).

dr7ana replied:
I'm using a different `get_timestamp` that takes a template for explicit type safety. Also, the only difference between `get_time` and `get_timestamp` is `.time_since_epoch()`
```c++
std::chrono::steady_clock::time_point get_time()
{
    return std::chrono::steady_clock::now();
}
std::chrono::nanoseconds get_timestamp()
{
    return std::chrono::steady_clock::now().time_since_epoch();
}
```

jagerman replied:
> Also, the only difference between get_time and get_timestamp is .time_since_epoch()

That *is* the important difference.  The latter is an abitrary duration that is meaningless on its own; the former is a object that is a duration with an epoch that fully defines it.

dr7ana replied:
Okay so `get_time` is an arbitrary duration that is meaningless on its own? So makes sense why we shouldn't use that one

dr7ana replied:
I think if you write this out you will see that it's effectively the same

jagerman replied:
No, other way around.  I had former/latter backwards in my original comment.

dr7ana replied:
If you can write this out inline and show me it's simpler and an improvement I will happily make the change, but I don't see your argument on this

jagerman replied:
> I think if you write this out you will see that it's effectively the same

Effectively the same in its final value, but different in its semantic meaning.

You can write all sorts of things that are effectively the same, but the use of types (i.e. abstractions) is to give semantic meaning through your intermediate types of what it is that you hold in your hand.  std::chrono::nanosecond is, under the hood, just an integer, but the type adds meaning to that integer.

A std::chrono::time_point is, under the hood, just a duration with an implied reference point, but that still matters, because nanoseconds with a reference point has a different semantic meaning than nanoseconds alone.

dr7ana replied:
Except that taking an `std::chrono` type (milliseconds, nanoseconds, etc) and duration casting it is the simpler way to do it. Taking a time point requires either appending `.time_since_epoch()` (which makes it *exactly* like `get_timestamp()`) or finding the difference between it and another time_point, at which point one must duration cast...

jagerman replied:
I don't know why the comment I wrote on this yesterday isn't here, so posting it again:

The principle here is that you don't throw away type safety before you need to without good reason: which means you should keep time points as time point objects, not durations which lose some semantic meaning.

The code here to keep time points as time points is not any more complicated (and only involves one duration cast instead of two), while not throwing away the type safety in intermediate steps:
```C++
                call_soon([this,
                           func = std::move(hook),
                           target_time = get_time() + delay]() mutable {
                    if (auto now = get_time(); now >= target_time)
                        func();
                    else
                        add_oneshot_event(
                                std::chrono::duration_cast<std::chrono::microseconds>(target_time - now),
                                std::move(func));
                });
```

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -124,7 +183,75 @@ namespace oxen::quic
+                    if (updated_delay <= 0us)
+                        func();
+                    else
+                        add_oneshot_event(updated_delay, std::forward<Callable>(func));
```
See other comment about a harmful inner forward.

jagerman commented on include/oxen/quic/loop.hpp:
```diff
@@ -124,7 +183,75 @@ namespace oxen::quic
+        {
+            {
+                std::lock_guard lock{job_queue_mutex};
+                job_queue.emplace(std::forward<Callable>(f));
```
Making this `forward` is counterproductive because it doesn't save anything: because job_queue contains `std::function<void()>`, there's no advantage in forwarding the reference.  All that happens is we carry a reference rather than a value, but then we hit the std::function constructor in the emplace, versus taking this as a `std::function<void()>` in the first place and just moving into the job_queue.

That would be more or less harmless, but because it's being converted from an untemplated function into a generic templated function, that means that pretty much every `call_soon` ends up compiling and either linking or inlining its own unique instantiation of the body of `call_soon` (because every lambda is a different `Callable` type).

jagerman commented on src/loop.cpp:
```diff
@@ -29,6 +29,111 @@ namespace oxen::quic
+        return timeval{
+                .tv_sec = static_cast<decltype(timeval::tv_sec)>(t / 1s),
+                .tv_usec = static_cast<decltype(timeval::tv_usec)>((t % 1s) / 1us)};
+    }
```
I don't understand the comment/code: on both platforms it appears that `tv_sec` is a `time_t` and `tv_usec` is a `suseconds_t` (regardless of the intermediate internal typedef name), so it ought to suffice to `static_cast<time_t>` and `static_cast<suseconds_t>` here.

dr7ana replied:
I explicitly specified this to fix an issue in CI, you can change it back and see how it will fail the mac builds

dr7ana replied:
There are tons of issues online documenting [this](https://github.com/szymonwieloch/rust-rawsock/issues/5)

jagerman replied:
macos seems happy with this:
```bash
om2g:~$ cat timeval.cpp 
```
```C++
#include <sys/time.h>
#include <chrono>
#include <iostream>

using namespace std::literals;

timeval loop_time_to_timeval(std::chrono::microseconds t)
{
    return timeval{
            .tv_sec = static_cast<time_t>(t / 1s),
            .tv_usec = static_cast<suseconds_t>((t % 1s) / 1us)};
}

int main() {
    auto x = loop_time_to_timeval(123456ms);
    std::cout << x.tv_sec << "s, " << x.tv_usec << "µs\n";
}
```
```bash
om2g:~$ g++ -std=c++20 timeval.cpp -o timeval
om2g:~$ ./timeval 
123s, 456000µs
```

Do you have a CI job number where this was failing?

dr7ana replied:
Can you explain why that is an improvement over the current cast to whatever the system defined `timeval` struct is? The above seems insufficient for other system implementations that can [compile things away](https://codebrowser.dev/gcc/include/x86_64-linux-gnu/bits/types/struct_timeval.h.html), like 64-bit `__time64_t`

jagerman replied:
Because `timeval` is a [documented, decades old POSIX standard](https://pubs.opengroup.org/onlinepubs/7908799/xsh/systime.h.html) that is required to have a `time_t tv_sec;` and `suseconds_t tv_usec;`.  Whatever `__time64_t` or `__darwin_suseconds_t` get used by standard library implementations is irrelevant: those are internal implementation types which are required to be the same as `time_t` and `suseconds_t` typedef in order to satify the `timeval` specification.

dr7ana replied:
Except `time_t` is *itself* a [typedef](https://en.wikipedia.org/w/index.php?title=Time_t&oldid=450752800):

>This type is a [typedef](https://en.wikipedia.org/wiki/Typedef) defined in the standard [<time.h>](https://en.wikipedia.org/wiki/Time.h) header. [ISO C](https://en.wikipedia.org/wiki/ISO_C) defines time_t as an arithmetic type, but does not specify any particular [type](https://en.wikipedia.org/wiki/Data_type), [range](https://en.wikipedia.org/wiki/Range_(mathematics)), [resolution](https://en.wikipedia.org/wiki/Frequency), or encoding for it. Also unspecified are the meanings of arithmetic operations applied to time values.

Casting to the native timeval types is the straight-line, most efficient solution to this.

jagerman replied:
> Except time_t is itself a typedef

Yup.  It's a typedef that is defined to be the type that goes into a `tv_sec`, and `suseconds_t` is likewise a typedef that is defined to be the type that goes into `tv_usec`.

dr7ana replied:
That is also incorrect. Please feel free to [verify this yourself](https://en.cppreference.com/w/c/chrono/time_t). For extra thoroughness, please see the latest [C17 standard](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf), specifically section 7.27.1-4 to see:

> The range and precision of times representable in clock_t and time_t are implementation-defined

`time_t` is not typedef'ed specifically for `tv_sec`. It is used in a whole host of other functions that can be enumerated by searching "time_t" on the C17 standard pdf file.

On my system, `time_t` is a typedef for `__time_t`, which *itself* typedefs a few times all the way down to `long int` on my system

dr7ana replied:
That is also incorrect, it is not typedef'ed specifically for that use-case. Please feel free to [verify this yourself](https://en.cppreference.com/w/c/chrono/time_t)
> Although not defined by the C standard, this is almost always an integral value holding the number of seconds

For reference to what they're speaking about, please see the latest [C17 standard](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf), specifically section 7.27.1-4 to see:

> The range and precision of times representable in clock_t and time_t are implementation-defined

`time_t` is not typedef'ed specifically for `tv_sec`. It is used in a whole host of other functions that can be enumerated by searching "time_t" on the C17 standard pdf file.

On my system, `time_t` is a typedef for `__time_t`, which *itself* typedefs a few times all the way down to `long int` on my system

jagerman replied:
Not resolved.

Again: `timeval` is explicitly defined in terms of `time_t` and `suseconds_t`, and thus *those are the types* that belong here.  Not `__time_t`, not `decltype(...)`, not `long int`, not `__darwin_time_t`, but simply the member types that the timeval API explicitly defines, which are `time_t` and `suseconds_t`.

dr7ana replied:
Well hasn't this certainly been an educational experience. In the interest of not continuing to block this for a third week over named booleans and other nit-picks, it is changed to read as you want it

dr7ana replied:
[Beautiful](https://ci.oxen.rocks/oxen-io/oxen-libquic/1115/23/3)

dr7ana replied:
[Well](https://ci.oxen.rocks/oxen-io/oxen-libquic/1116/23/3), look like we can securely resolve this now.

jagerman commented on src/loop.cpp:
```diff
@@ -29,6 +29,111 @@ namespace oxen::quic
+            return false;
+        }
+
+        ev.reset();
+        f = nullptr;
```
This doesn't seem thread-safe: someone could call `stop()` from any thread which means the clearing of the existing `f` could happen while `f` is currently in the middle of a call in the event loop.

(This becomes irrelevant if we get rid of `stop()` in favour of using the destructor, which *is* thread-safe because destruction gets deferred to the event loop).

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+        const int NUM_ITERATIONS{10};
+        constexpr auto msg = "hello from the other siiiii-iiiiide"_bsv;
+
+        std::promise<bool> d_promise;
```
The `bool` value that goes into this promise never varies and is never read; this should be a `std::promise<void>` instead.

jagerman replied:
Actually, given where this is used below, this should stay a `promise<bool>` or maybe a `promise<std::pair<bool, bool>>`, and the conditions currently in the REQUIRE should get propagated via the promise's value to be checked via REQUIRE's outside the call_soon.

dr7ana replied:
See above response to the `REQUIRE` comment

jagerman replied:
Not resolved: if this `bool` value is not being used, this should be a `promise<void>`.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+
+        test_net.call_later(1s, [&]() {
+            REQUIRE(recv_counter == send_counter);
+            REQUIRE(!life);
```
Invoking REQUIRE in another thread like this is a recipe for crashes with catch2; you *can* make it work if you use a mutex to guard anything that touches catch2, but it would be easier to get these values out via a promise and then REQUIRE after the promise gets fulfilled.

jagerman replied:
(And there's a `d_promise` just below this that can do the job)

dr7ana replied:
I understand your motivation behind this. Few thoughts on that:
- By the time the `::call_later()` is invoked, the sending and receiving has stopped
- The counters are atomic
- There are literally tons of other non-problematic `REQUIRE` checks in the test code that aren't guarded by mutexes

jagerman replied:
Not resolved.

`REQUIRE` inside a callback is something we should avoid if easy to do so precisely because of the Catch2 thread hazard; even if this call does not hit the problem, it is also extremely simple to move the `REQUIRE`s after the `require_future` without sacrificing anything in terms of the test or clarity, but gaining that it is immediately obviously safe since there's no longer any catch2 interactions happening on different threads in the test.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+        REQUIRE(handler->is_running());
+
+        test_net.call_later(1s, [&]() {
+            REQUIRE(recv_counter == send_counter);
+            prom_a.set_value(true);
```
```suggestion
            prom_a.set_value(recv_counter == send_counter);
```

and then use REQUIRE on the value returned from the future.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+        REQUIRE(handler->start());
+
+        test_net.call_later(1s, [&]() {
+            REQUIRE(recv_counter == send_counter);
+            prom_b.set_value(true);
```
As commented above.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+
+        REQUIRE(handler->is_running());
+
+        test_net.call_later(1s, [&]() {
```
1s seems too long here, given the 10ms repeat + 10 iterations.  I'd think these test should be able to complete in `2 * 10ms * NUM_ITERATIONS` (200ms; doubled from the ideal 100ms for an error margin), and if they are taking longer than that (but less than 1s) it likely indicates a problem with the event scheduling.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,148 @@
+
+        REQUIRE(handler->start());
+
+        test_net.call_later(1s, [&]() {
```
Too long, see above.

jagerman commented on src/loop.cpp:
```diff
@@ -29,6 +29,113 @@ namespace oxen::quic
+    {
+        clear_old_tickers();
+        auto t = make_shared<Ticker>();
+        _tickers.push_back(t->weak_from_this());
```
```suggestion
        _tickers.push_back(t);
```
But also related, is there a reason `Ticker` needs the added complexity of inheriting from enable_shared_from_this?

dr7ana replied:
It's not a lot of overhead, and it simplifies. I don't really see how this is more complex, could you explain?

jagerman replied:
What does it simplify?

The complexity is in design: rather than being a simple class without inheritance, it's an inheriting class adding a capability that isn't needed or used anywhere.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,144 @@
+
+namespace oxen::quic::test
+{
+    struct lifetime : public std::enable_shared_from_this<lifetime>
```
Why enable_shared_from_this?

dr7ana replied:
Above comment

jagerman replied:
Not resolved.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,144 @@
+                life.reset();
+        });
+
+        test_net.call_later(1s, [&]() {
```
Too long: this should match the duration in the other test.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,144 @@
+        constexpr auto msg = "hello from the other siiiii-iiiiide"_bsv;
+
+        std::promise<bool> prom_a, prom_b;
+        std::future<bool> fut_a = prom_a.get_future(), fut_b = prom_b.get_future();
```
Unused/unneeded `bool` here; these should be `promise<void>`/`future<void>`.

jagerman commented on tests/013-eventhandler.cpp:
```diff
@@ -0,0 +1,144 @@
+                // else
+                // {
+                //     handler->stop();
+                // }
```
Dead code?

dr7ana committed:
Review fixes

- Minor changes; renaming, test suite tweaks
- timeval casting and get_time() in ::call_later RE: Jason comments

dr7ana committed:
correct casting

dr7ana commented:
@jagerman ready for merge pending any final comments/thoughts/concerns
